### ISSUE TITLE
[Feature] Wanda++ for LM head pruning

### ISSUE THREAD

**Issue by namgyu-youn** [pytorch/ao#2517]
### Background
Current Wanda (`torchao/sparsity/wanda.py`) uses magnitude-based pruning criteria: `|weight| * ||activation||` ([arXiv](https://arxiv.org/abs/2306.11695), fig1)

<img width="600" height="300" alt="Image" src="https://github.com/user-attachments/assets/b660d9a5-44b7-44db-88c7-e5d3af77d877" />

### Feature : Wanda++

And recently, Wanda++ ([arXiv](https://arxiv.org/abs/2503.04992)) was published. The main difference can be summarized like the following:
- **Regional Gradient Score (RGS)**: Uses block-level gradients instead of magnitude
- **Regional Optimization (RO)**: Block-level weight fine-tuning after pruning (iterately)

How about expanding Wanda into Wanda++? I am not certain if I can handle full RO algorithm, but Table 1 shows significant improvements even with RGS.

<img width="769" height="465" alt="Image" src="https://github.com/user-attachments/assets/b8186195-c2e6-4325-86b0-7034a54ce59b" />

**Comment by namgyu-youn** [comment]
@drisspg @jcaip Hi, can I take this issue? I'm deeply interested in implementing the Regional Gradient Score (RGS) from Wanda++. 

And since Regional Optimization (RO) is the fine-tuning concept, there will be a chance for RO after applying RGS.

**Comment by jcaip** [comment]
@namgyu-youn This is cool, but I'm not sure that I want torchao to be a repository for all the different pruning methods.

My reasoning for this is primarily on the maintenance side - accuracy techniques are something we've historically left up to researchers while we've focused on acceleration (take sparse model, make it faster). I dont have bandwidth to keep up to date with all the latest papers and add them to torchao myself. And we would also force them to use our abstraction, which I don't think covers all use cases. 

If you want to make your case for why Wanda++ in particular is important though, I'll hear you out.

**Comment by namgyu-youn** [comment]
> My reasoning for this is primarily on the maintenance side - accuracy techniques are something we've historically left up to researchers while we've focused on acceleration (take sparse model, make it faster). I dont have bandwidth to keep up to date with all the latest papers and add them to torchao myself. And we would also force them to use our abstraction, which I don't think covers all use cases.

@jcaip  Same thought I felt while reviewing papers related to sparsity. The weird point is, most of the researches are not reproducible, requiring developers too much cost as you mentioned; whatever they zeroize (weight, activation, etc) or whatever criteria is or whatever accuracy they scored.

Therefore, I agree with pausing for more pruning papers, but I hope Wanda++ would be a adopted because: 
1. Wanda is one of the well-used techniques across GitHub and HF; https://huggingface.co/models?sort=trending&search=wanda
2. Since Wanda++ is a future work of Wanda, they share many logics. I tried first implemtation at https://github.com/namgyu-youn/ao/blob/855b5440523d00b6ff4ea6f82e103b73a4065720/torchao/prototype/sparsity/pruner/wanda_pp.py and since the difference is magnitude vs. gradient, the implementation can be a quiet easy

**Comment by jcaip** [comment]
@namgyu-youn

It still feels weird for me to make an exception for wanda++, because while 1) is true, wanda is far from dominant (and wanda ++ is not as popular as wanda). and for 2) just because it's easy to do doesn't mean that we should do it - if another pruning method is easy to add would we add it then?

Would you be interested in creating your own repo of pruning techniques + benchmarks and just using torchao as a dependency? I think it would be very valuable for the sparsity ecosystem and I'd be happy to link to it in the README / documentation. If that becomes popular (and if you don't want to keep it standalone) we can revisit adding it torchao then, as I think the value prop there is a bit clearer vs just a single wanda ++ implementation. 

Unfortunately for now I'm going to say no to adding a standalone wanda++ implementation to torchao though. 

If you're interested in other sparsity tasks though, please lmk - I'm still catching up from my leave but I can find some tasks for you :)

**Comment by namgyu-youn** [comment]
Fully open to benchmark and profiling, but because there is no plan to upstream, closing them seems right move to me. Let me know if there is anything I can help.
